xtc/src/common_types.rs

```diff
@@ -0,0 +1,117 @@
+use derive_builder::*;
+use derive_new::*;
+use ic_kit::{
+    candid::{CandidType, Int, Nat},
+    Principal,
+};
+use std::convert::{TryFrom, TryInto};
+use xtc_history_common::types::*;
+
+type Time = Int;
+
+#[derive(CandidType, Clone)]
+pub enum Operation {
+    approve,
+    mint,
+    transfer,
+    transferFrom,
+}
+
+#[derive(CandidType)]
+pub struct Metadata<'a> {
+    pub decimals: u8,
+    pub fee: Nat,
+    pub logo: &'a str,
+    pub name: &'a str,
+    pub owner: Principal,
+    pub symbol: &'a str,
+    pub totalSupply: Nat,
+}
+
+#[derive(CandidType, Debug, Eq, PartialEq)]
+pub enum TxError {
+    InsufficientAllowance,
+    InsufficientBalance,
+}
+
+pub type TxReceipt = Result<Nat, TxError>;
+
+#[derive(CandidType, Clone, new)]
+pub struct TxRecord {
+    pub caller: Option<Principal>,
+    pub from: Principal,
+    pub to: Principal,
+    pub amount: Nat,
+    pub fee: Nat,
+    pub op: Operation,
+    pub timestamp: Time,
+    pub index: Nat,
+}
+
+impl TxRecord {
+    pub fn setIndex(mut self, index: Nat) -> Self {
+        self.index = index;
+        self
+    }
+}
+
+impl TryFrom<Transaction> for TxRecord {
+    type Error = ();
+    fn try_from(transaction: Transaction) -> Result<TxRecord, ()> {
+        let mut tx_record: TxRecord;
+        match transaction.kind {
```

Comment:
I was thinking about this yesterday night, the standard only support 4 + 1 events: approve, transfer, transferFrom, mint, burn, but XTC has many more. If any clients like plug needs those events, we will need to fit them in here. @rocky-fleek what events Plug rely on? I suppose it has a transaction history which shows all kind of extra events like canister calls...

Reply:
I think it is only transfer/mint/burn but **not** canister created and canister called.

Reply:
when surfacing data for the ERC-20 standard you can convert a canisterCreated and canisterCalled events to *burn*.

Reply:
Plug needs the following methods to work as it is right now:
```
'balance': (arg_0: [] | [Principal]) => Promise<bigint>,
  'burn': (arg_0: {
    canister_id: Principal,
    amount: bigint
  }) => Promise<BurnResult>
  'transfer': (
    arg_0: {
      'to': Principal,
      'from': [] | [Principal],
      'amount': bigint,
    },
  ) => Promise<TransferResult>,```